### PR TITLE
[FEAT #132] Don't show revert dialog if the file contents did not change

### LINKED ISSUE #132: Improve revert detection: don't show a dialog if nothing changed
Previous versions of SubEthaEdit were very conservative with their revert behavior. Any change of file (timestamp only too) would lead to a prompt. I think in these modern times, where the likelyhood of e.g. git just replacing the whole file and putting it back again because of branch switches etc is high, we should be more forgiving.

Explore only showing the revert dialog if the content of the file has changed.

Related to: #125

### PR DESCRIPTION
**Pull request by fourplusone**
This PR resolves #132

### CODE REVIEW

File: SubEthaEdit-Mac/Source/PlainTextDocument.h [hunk 1]
```diff
@@ -89,6 +89,9 @@ extern NSString * const PlainTextDocumentDidSaveShouldReloadWebPreviewNotificati
     DocumentMode  *I_documentMode;
     FoldableTextStorage *I_textStorage;
     
+    // I_fileHash represents the MD5 hash of the document
+    NSData *I_fileHash;
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.h]: style: please put new ivars if possible in the implementation side, also switch to regular _ prefixes as we'd like to get away from the `I_`, `O_` etc. old school prefixes.

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 2]
```diff
@@ -2390,7 +2390,20 @@ - (void)TCM_ensureFileTypeDataOrSEEText {
 - (void)saveDocumentWithDelegate:(id)delegate didSaveSelector:(SEL)didSaveSelector contextInfo:(void *)contextInfo {
     if ([self TCM_validateDocument]) {
 		[self TCM_ensureFileTypeDataOrSEEText];
-        [super saveDocumentWithDelegate:delegate didSaveSelector:didSaveSelector contextInfo:contextInfo];
+        
+        // If the file exists TCM_validateDocument handles the case whether a file
+        // has been modified by another application, so the super method would display
+        // a misleading dialog.
+        if([NSFileManager.defaultManager fileExistsAtPath:self.fileURL.path]) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: Not sure if this redirection is correct this way. (Also we should have called the didSaveSelector even on cancellation in the past too if I read the docs right). Need to look at the behaviour you like to circumvent. 

Also style: please space after `if` and before `(`

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 3]
```diff
@@ -2390,7 +2390,20 @@ - (void)TCM_ensureFileTypeDataOrSEEText {
 - (void)saveDocumentWithDelegate:(id)delegate didSaveSelector:(SEL)didSaveSelector contextInfo:(void *)contextInfo {
     if ([self TCM_validateDocument]) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: style: please do reindent the whole function with ctrl-I if we leading whitespace on some of the lines.

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 4]
```diff
@@ -2680,7 +2693,14 @@ - (BOOL)readSEETextFromURL:(NSURL *)anURL properties:(NSDictionary *)aProperties
     [[self session] addContributors:contributors];
     // load text into dictionary
     NSMutableDictionary *storageRep = [dictRep objectForKey:@"TextStorage"];
-    NSString *string = [NSString stringWithContentsOfURL:[NSURL fileURLWithPath:[[anURL path] stringByAppendingPathComponent:@"plain.txt"]]  encoding:[dictRep objectForKey:@"AutosaveInformation"] ? NSUTF8StringEncoding : (NSStringEncoding)[[storageRep objectForKey:@"Encoding"] unsignedIntValue] error:outError];
+    
+    NSURL *textDataURL = [NSURL fileURLWithPath:[[anURL path] stringByAppendingPathComponent:@"plain.txt"]];
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: style: on rearranging please also switch to property accessors e.g. `anURL.path` in this case

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 5]
```diff
@@ -2680,7 +2693,14 @@ - (BOOL)readSEETextFromURL:(NSURL *)anURL properties:(NSDictionary *)aProperties
...
+    NSData *textData = [NSData dataWithContentsOfURL:textDataURL options:0 error:outError];
+    
+    if(!textData) return NO;
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: style: space after `if`, and please no new if/else bodyes without curlys. prefarably on extra line too. e.g. 

```objective-c
if (!textData) {
  return NO;
}
```

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 6]
```diff
@@ -2680,7 +2693,14 @@ - (BOOL)readSEETextFromURL:(NSURL *)anURL properties:(NSDictionary *)aProperties
...
+    
+    NSString *string = [NSString stringWithData:textData encoding:[dictRep objectForKey:@"AutosaveInformation"] ? NSUTF8StringEncoding : (NSStringEncoding)[[storageRep objectForKey:@"Encoding"] unsignedIntValue]];
+    
     if (!string) return NO;
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: dito

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 7]
```diff
@@ -3787,13 +3844,50 @@ - (NSData *)dataOfType:(NSString *)aType error:(NSError **)outError {
...
+        return NO;
+    }
+    
+    // If the file size is different, the contents also differ for sure
+    if (fattrs.fileSize != self.fileAttributes.fileSize) {
+        return YES;
+    }
+    
+    NSData *fileData;
+    if(I_flags.isSEEText) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: style: space

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 8]
```diff
@@ -3787,13 +3844,50 @@ - (NSData *)dataOfType:(NSString *)aType error:(NSError **)outError {
...
 }
 
+- (BOOL)onDiskRepresentationHasChanged {
+    NSURL *fileURL = self.fileURL;
+    NSString *fileName = [fileURL path];
+    
+    NSDictionary *fattrs = [[NSFileManager defaultManager] attributesOfItemAtPath:fileName error:nil];
+    
+    if (!fattrs) {
+        return YES;
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: isn't this also unhappy for new files?

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 9]
```diff
@@ -3787,13 +3844,50 @@ - (NSData *)dataOfType:(NSString *)aType error:(NSError **)outError {
...
+    }
+    
+    // If the modificationDate has not changed we can assume, that the content is still the same
+    if ([[fattrs fileModificationDate] compare:[[self fileAttributes] fileModificationDate]] == NSOrderedSame) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: style: please use property accessors for new code

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 10]
```diff
@@ -3787,13 +3844,50 @@ - (NSData *)dataOfType:(NSString *)aType error:(NSError **)outError {
...
+
 - (BOOL)TCM_validateDocument {
     NSString *fileName = [[self fileURL] path];
     DEBUGLOG(@"FileIOLogDomain", SimpleLogLevel, @"Validate document: %@", fileName);
 
-    NSDictionary *fattrs = [[NSFileManager defaultManager] attributesOfItemAtPath:fileName error:nil];
-
-    if ([[fattrs fileModificationDate] compare:[[self fileAttributes] fileModificationDate]] == NSOrderedSame) {
+    
+    if (![self onDiskRepresentationHasChanged] || ![NSFileManager.defaultManager fileExistsAtPath:fileName]) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: The order of the two items in the if probably should change if we treat a non-existing file as no change anyways.

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 11]
```diff
@@ -3578,6 +3617,7 @@ - (BOOL)writeToURL:(NSURL *)absoluteURL ofType:(NSString *)inType forSaveOperati
 		}
 		//        NSArray *xattrKeys = [UKXattrMetadataStore allKeysAtPath:[absoluteURL path] traverseLink:YES];
 		//        NSLog(@"%s xattrKeys:%@",__FUNCTION__,xattrKeys);
+        I_fileHash = [writtenData md5Data];
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: Do we want to update this for all the cases here? E.g. autosaving should not update it, save as should not either. Those things should be conveyed through the save operation type if I remember correctly.

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 12]
```diff
@@ -3549,21 +3576,33 @@ - (BOOL)writeToURL:(NSURL *)absoluteURL ofType:(NSString *)inType forSaveOperati
...
+            } else {
+                writtenData = nil;
             }
         } else {
             // let us write using NSStrings write methods so the encoding is added to the extended attributes
-            result = [[[(FoldableTextStorage *)[self textStorage] fullTextStorage] string] writeToURL:absoluteURL atomically:NO encoding:[self fileEncoding] error:outError];
+            NSString *text = [[(FoldableTextStorage *)[self textStorage] fullTextStorage] string];
+            NSStringEncoding encoding = [self fileEncoding];
+            
+            result = [text writeToURL:absoluteURL atomically:NO encoding:encoding error:outError];
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: This one makes me a little bit uneasy. We only ose the string method to also write the system wide xattrs it writes. Maybe we should handle the com.apple.TextEncoding safe always on ourn own and make sure the NSData really is the thing we wrote.

File: SubEthaEdit-Mac/Source/PlainTextDocument.m [hunk 13]
```diff
@@ -3549,21 +3576,33 @@ - (BOOL)writeToURL:(NSURL *)absoluteURL ofType:(NSString *)inType forSaveOperati
...
     DEBUGLOG(@"FileIOLogDomain", AllLogLevel, @"write to:%@ type:%@ saveOperation:%lu originalURL:%@", absoluteURL, inType, (unsigned long)saveOperation,originalContentsURL);
+    
+    NSData *writtenData;
+    
     if (UTTypeConformsTo((__bridge CFStringRef)inType, kUTTypeData)) {
         BOOL modeWantsUTF8BOM = [[[self documentMode] defaultForKey:DocumentModeUTF8BOMPreferenceKey] boolValue];
         DEBUGLOG(@"FileIOLogDomain", SimpleLogLevel, @"modeWantsUTF8BOM: %d, hasUTF8BOM: %d", modeWantsUTF8BOM, I_flags.hasUTF8BOM);
         BOOL useUTF8Encoding = ((I_lastSaveOperation == NSSaveToOperation) && (I_encodingFromLastRunSaveToOperation == NSUTF8StringEncoding)) || ((I_lastSaveOperation != NSSaveToOperation) && ([self fileEncoding] == NSUTF8StringEncoding));
 		BOOL result = NO;
         if ((I_flags.hasUTF8BOM || modeWantsUTF8BOM) && useUTF8Encoding) {
```
Comment (monkeydom) [inline SubEthaEdit-Mac/Source/PlainTextDocument.m]: Just as a sidenote: We really should treat UTF8 with BOM as a separate encoding. This mode setting + regular UTF8 Encoding is quite messy.

**Comment by fourplusone** [comment]
I did some improvements

- [x] network drives and behaviour there: **For now, network files won't be reread**
- [ ] what happens with files that we don't have regular read access to, but admin account dialoged our way to?
- [x] reading large files? **Files larger than 1MB won't be reread**

**Comment by monkeydom** [comment]
I think we should just try to put that in and see if this works for us in daily usage.

**Comment by monkeydom** [comment]
A build with this in is at https://subethaedit.net/Releases/SubEthaEdit-5.1.2a-9651.zip

**Pull request merged**